### PR TITLE
rpc server: break legacy CLI options and remove "backward compatible HTTP server"

### PR DESCRIPTION
**Pull request by niklasad1**
Close https://github.com/paritytech/substrate/issues/13049 https://github.com/paritytech/substrate/issues/13561

The CLI options that are removed/or renamed are:

1.  --rpc--max--payload (replaced by --rpc--max-request-size and --rpc-max-response-size)
3. --ws--max--out-buffer-capacity
4. --ws-external (replaced by --rpc_external)
5. --unsafe--ws--external (replaced by --unsafe-rpc-external)
6. --ipc--path 
7. --ws-port (replaced by --rpc--port)
8. --ws-max--connections (replaced by --rpc--max--connections)
9. --rpc-http (replaced by --rpc-addr)
10. --rpc-ws (replaced by --rpc-addr)

polkadot companion: https://github.com/paritytech/polkadot/pull/6987

cumulus companion: https://github.com/paritytech/cumulus/pull/2417

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2388058

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2388059

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2388057

**Comment by niklasad1** [comment]
not stale

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551778

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551783

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551776

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551777

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2551787

File: client/service/src/lib.rs [hunk 1]
```diff
@@ -380,57 +380,35 @@ where
...
-	match tokio::task::block_in_place(|| {
-		config.tokio_handle.block_on(futures::future::try_join(http_fut, ws_fut))
-	}) {
-		Ok((http, ws)) => Ok(Box::new((waiting::Server(Some(http)), waiting::Server(Some(ws))))),
+	let server_fut = sc_rpc_server::start_server(server_config);
+
+	match tokio::task::block_in_place(|| config.tokio_handle.block_on(server_fut)) {
+		Ok(server) => Ok(Box::new(waiting::Server(Some(server)))),
 		Err(e) => Err(Error::Application(e)),
 	}
```
Comment (bkchr) [inline client/service/src/lib.rs:413]:
```suggestion
	match config.tokio_handle.block_on(server_fut) {
		Ok(server) => Ok(Box::new(waiting::Server(Some(server)))),
		Err(e) => Err(Error::Application(e)),
	}
```

We don't really need `block_in_place` here? Maybe we should also make the entire function async

File: client/service/src/lib.rs [hunk 2]
```diff
@@ -380,57 +380,35 @@ where
...
+	let addr = config
+		.rpc_addr
 		.unwrap_or_else(|| "127.0.0.1:9944".parse().expect("valid sockaddr; qed"));
-	let ws_addr2 = random_port(ws_addr);
-
-	let http_addr = config
-		.rpc_http
-		.unwrap_or_else(|| "127.0.0.1:9933".parse().expect("valid sockaddr; qed"));
-	let http_addr2 = random_port(http_addr);
+	let addr2 = random_port(addr);
```
Comment (bkchr) [inline client/service/src/lib.rs:391]: I assume this is the "backup" address if the requested port is blocked? Some comment would be nice.

File: client/cli/src/commands/run_cmd.rs [hunk 3]
```diff
@@ -122,25 +102,13 @@ pub struct RunCmd {
...
-	/// Specify HTTP RPC server TCP port.
+	/// Specify JSON-RPC server TCP port.
 	#[arg(long, value_name = "PORT")]
 	pub rpc_port: Option<u16>,
 
-	/// Specify WebSockets RPC server TCP port.
-	#[arg(long, value_name = "PORT")]
-	pub ws_port: Option<u16>,
-
 	/// Maximum number of WS RPC server connections.
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply client/cli/src/commands/run_cmd.rs:109]: thanks, you are right

File: client/service/src/lib.rs (hunk 1 above)
Comment (niklasad1) [reply client/service/src/lib.rs:413]: I like the idea of making the entire function async but IIRC I didn't want to starve other spawned tasks but given that this is on startup it is probably fine.

I'll have look

Comment (bkchr) [reply client/service/src/lib.rs:412]: Yeah we would just move the block_on call. I think we should block to ensure that we bring down the node if the RPC server failed to start.

So, TLDR: Maybe the best is to only apply my suggested changes.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2620452

File: client/service/src/lib.rs (hunk 1 above)
Comment (niklasad1) [reply client/service/src/lib.rs]: added a comment, we need it because some tests calls this function and already blocks on it in.

**Review by koute — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]

File: client/rpc-servers/src/lib.rs [hunk 4]
```diff
@@ -175,14 +130,28 @@ pub async fn start<M: Send + Sync + 'static>(
...
 }
 
+fn hosts_filtering(enabled: bool, addrs: &[SocketAddr]) -> AllowHosts {
+	if enabled {
+		// NOTE The listening addresses are whitelisted by default.
+		let mut hosts = Vec::with_capacity(addrs.len() * 2);
+		for addr in addrs {
+			hosts.push(format!("localhost:{}", addr.port()).into());
+			hosts.push(format!("127.0.0.1:{}", addr.port()).into());
+		}
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply client/rpc-servers/src/lib.rs:148]: sure why not

[the remark this answers is not in the document]
Comment (niklasad1) [reply client/rpc-servers/src/lib.rs]: @dmitry-markin can you double check my latest commit? Not really that familiar with the IPv6 addresses

[the remark this answers is not in the document]
Comment (niklasad1) [reply client/rpc-servers/src/lib.rs]: never mind, our parsing of host address doesn't work for this anyway.

lemme create an issue in jsonrpsee for that.

**Comment by pepoviola** [comment]
Hi @niklasad1, this require some changes in zombienet. I'm tracking those [here](https://github.com/paritytech/zombienet/issues/924)

**Comment by niklasad1** [comment]
bot rebase

File: client/rpc-servers/src/lib.rs [hunk 5]
```diff
@@ -64,98 +66,51 @@ pub struct WsConfig {
...
 		.layer(try_into_cors(cors)?);
 
 	let mut builder = ServerBuilder::new()
-		.max_request_body_size(max_payload_in)
-		.max_response_body_size(max_payload_out)
-		.max_connections(max_connections)
-		.max_subscriptions_per_connection(max_subs_per_conn)
+		.max_request_body_size(payload_size_or_default(max_payload_in_mb))
+		.max_response_body_size(payload_size_or_default(max_payload_out_mb))
+		.max_connections(max_connections.map_or(RPC_MAX_CONNECTIONS, |m| m as u32))
```
Comment (melekes) [inline client/rpc-servers/src/lib.rs:106]: this is confusing.. we say earlier that "default is None (unlimited)" number of connections, but here we seem to restrict the number of connections to 100

Comment (niklasad1) [reply client/rpc-servers/src/lib.rs:106]: Earlier we did only `restrict ws connections to 100 by default` and the http server had "unlimited allowed connections" 

After this PR there is not a separate ws and http server so I thought it was most reasonable to restrict it to 100 by default but I understand your point.

Do you prefer something else?

Comment (melekes) [reply client/rpc-servers/src/lib.rs:106]: Should we update the docs to reflect the new default (100)?

Comment (niklasad1) [reply client/rpc-servers/src/lib.rs:106]: Sounds good to me.

What do you reckon changing the `structopt CLI Option<usize>` to `usize` and provide a default value?
Probably the easiest to maintain in the long run

Comment (melekes) [reply client/rpc-servers/src/lib.rs:106]: > What do you think about changing the structopt CLI Option<usize> to usize and provide a default value?
Then we don't really have be bother unwrap_or in the rpc code itself.

I'm in favor

**Review by alvicsam — APPROVED** [review]

**Review by sergejparity — APPROVED** [review]

**Comment by niklasad1** [comment]
@pepoviola I bumped zombienet to v1.3.47 because v1.3.43 wasn't found on dockerhub but maybe that doesn't work...

**Comment by niklasad1** [comment]
bot merge

**Pull request merged**